SDWebImage/UIView+WebCache.m

```diff
@@ -54,28 +67,48 @@ - (void)sd_internalSetImageWithURL:(nullable NSURL *)url
         id <SDWebImageOperation> operation = [SDWebImageManager.sharedManager loadImageWithURL:url options:options progress:progressBlock completed:^(UIImage *image, NSData *data, NSError *error, SDImageCacheType cacheType, BOOL finished, NSURL *imageURL) {
             __strong __typeof (wself) sself = wself;
             [sself sd_removeActivityIndicator];
-            if (!sself) {
-                return;
-            }
-            dispatch_main_async_safe(^{
-                if (!sself) {
-                    return;
-                }
-                if (image && (options & SDWebImageAvoidAutoSetImage) && completedBlock) {
-                    completedBlock(image, error, cacheType, url);
-                    return;
-                } else if (image) {
-                    [sself sd_setImage:image imageData:data basedOnClassOrViaCustomSetImageBlock:setImageBlock];
+            if (!sself) { return; }
+            BOOL shouldCallCompletedBlock = finished || (options & SDWebImageAvoidAutoSetImage);
```

Comment:
Maybe I always enter the trap. This case:

1. finished == NO
2. image != nil
3.  (options & SDWebImageAvoidAutoSetImage) == YES

The left one will not call completion but the right one will call. We should change it to
`BOOL shouldCallCompletedBlock = finished || (image && (options & SDWebImageAvoidAutoSetImage))`

Reply:
There must be a better way to express those conditions. Right now (on our branch) is a code smell. Easy to have missed cases like the one you noted above. Let's not settle with this and improve it more :)

Reply:
I've reviewed again those conditions and they are exactly as the code from 4.1.2
On your case (`finished=NO` + `image!=nil` + `(options & SDWebImageAvoidAutoSetImage) == YES`) the code from 4.1.2 was calling the completion and so is the code from this PR.

Reply:
@dreampiggy I think we are good for merge now.

Reply:
@bpoplauschi Well. It should be equal(The case I comment is OK. I omit the last completion call). If you think nothing to add, you can merge now.